File: packages/frontend/src/modules/opportunities/services/OpportunityService.js
// packages/frontend/src/modules/opportunities/services/OpportunityService.js
import api from '../../../services/api/apiClient';

/**
 * Service for managing opportunities
 */
class OpportunityService {
  /**
   * Fetch all opportunities
   * @returns {Promise} Promise resolving to opportunities data
   */
  getAllOpportunities() {
    console.log('Fetching all opportunities...');
    return api.get('/opportunities')
      .then(response => {
        console.log('Opportunities data:', response.data);
        return response;
      })
      .catch(error => {
        console.error('Error fetching opportunities:', error);
        throw error;
      });
  }

  /**
   * Fetch opportunity by ID
   * @param {string} id Opportunity ID
   * @returns {Promise} Promise resolving to opportunity data
   */
  getOpportunityById(id) {
    return api.get(`/opportunities/${id}`);
  }

  /**
   * Create a new opportunity
   * @param {Object} opportunity Opportunity data
   * @returns {Promise} Promise resolving to created opportunity
   */
  createOpportunity(opportunity) {
    return api.post('/opportunities', opportunity);
  }

  /**
   * Update an existing opportunity
   * @param {string} id Opportunity ID
   * @param {Object} opportunity Updated opportunity data
   * @returns {Promise} Promise resolving to updated opportunity
   */
  updateOpportunity(id, opportunity) {
    return api.put(`/opportunities/${id}`, opportunity);
  }

  /**
   * Delete an opportunity
   * @param {string} id Opportunity ID
   * @returns {Promise} Promise resolving to deletion result
   */
  deleteOpportunity(id) {
    return api.delete(`/opportunities/${id}`);
  }

  /**
   * Delete multiple opportunities
   * @param {Array} ids Array of opportunity IDs
   * @returns {Promise} Promise resolving to deletion result
   */
  deleteMultipleOpportunities(ids) {
    return api.delete('/opportunities/batch', { data: ids });
  }

  /**
   * Get opportunities by account name
   * @param {string} accountName Account name to filter by
   * @returns {Promise} Promise resolving to filtered opportunities
   */
  getOpportunitiesByAccountName(accountName) {
    return api.get(`/opportunities/account/${accountName}`);
  }

  /**
   * Get opportunities by stage
   * @param {string} stage Stage to filter by
   * @returns {Promise} Promise resolving to filtered opportunities
   */
  getOpportunitiesByStage(stage) {
    return api.get(`/opportunities/stage/${stage}`);
  }

  /**
   * Get opportunities by status
   * @param {string} status Status to filter by
   * @returns {Promise} Promise resolving to filtered opportunities
   */
  getOpportunitiesByStatus(status) {
    return api.get(`/opportunities/status/${status}`);
  }

  /**
   * Get opportunities by assignee
   * @param {string} assignee Assignee to filter by
   * @returns {Promise} Promise resolving to filtered opportunities
   */
  getOpportunitiesByAssignee(assignee) {
    return api.get(`/opportunities/assignee/${assignee}`);
  }

  /**
   * Search opportunities
   * @param {string} query Search query
   * @returns {Promise} Promise resolving to search results
   */
  searchOpportunities(query) {
    return api.get(`/opportunities/search?query=${query}`);
  }
}

export default new OpportunityService();

File: packages/backend/src/main/java/com/enterprise/modules/opportunity/service/OpportunityService.java
// packages/backend/src/main/java/com/crm/modules/opportunity/service/OpportunityService.java
package com.enterprise.modules.opportunity.service;

import com.enterprise.modules.opportunity.dto.OpportunityDto;
import java.util.List;

public interface OpportunityService {
    /**
     * Get all opportunities
     * @return List of all opportunities
     */
    List<OpportunityDto> getAllOpportunities();
    
    /**
     * Get an opportunity by ID
     * @param id Opportunity ID
     * @return Opportunity with the specified ID
     */
    OpportunityDto getOpportunityById(Long id);
    
    /**
     * Create a new opportunity
     * @param opportunityDto Opportunity data
     * @return Created opportunity with generated ID
     */
    OpportunityDto createOpportunity(OpportunityDto opportunityDto);
    
    /**
     * Update an existing opportunity
     * @param id Opportunity ID
     * @param opportunityDto Updated opportunity data
     * @return Updated opportunity
     */
    OpportunityDto updateOpportunity(Long id, OpportunityDto opportunityDto);
    
    /**
     * Delete an opportunity
     * @param id Opportunity ID
     * @return true if opportunity was deleted, false if not found
     */
    boolean deleteOpportunity(Long id);
    
    /**
     * Delete multiple opportunities
     * @param ids List of opportunity IDs to delete
     * @return Number of opportunities deleted
     */
    int deleteOpportunities(List<Long> ids);
    
    /**
     * Get opportunities filtered by account name
     * @param accountName Account name
     * @return List of opportunities with the specified account name
     */
    List<OpportunityDto> getOpportunitiesByAccountName(String accountName);
    
    /**
     * Get opportunities filtered by stage
     * @param stage Stage name
     * @return List of opportunities in the specified stage
     */
    List<OpportunityDto> getOpportunitiesByStage(String stage);
    
    /**
     * Get opportunities filtered by status
     * @param status Status (Open or Closed)
     * @return List of opportunities with the specified status
     */
    List<OpportunityDto> getOpportunitiesByStatus(String status);
    
    /**
     * Get opportunities filtered by assignee
     * @param assignee Assignee name
     * @return List of opportunities assigned to the specified assignee
     */
    List<OpportunityDto> getOpportunitiesByAssignee(String assignee);
    
    /**
     * Search opportunities by name, account name, or contacts name
     * @param query Search query
     * @return List of opportunities matching the search query
     */
    List<OpportunityDto> searchOpportunities(String query);
}